### PR TITLE
[Podchaser] Add extractor

### PR DESCRIPTION
**Pull request by connercsbn**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

This adds support for podchaser.com, a podcast aggregator. All the podcast info is grabbed in JSON from Podchaser's API (api.podchaser.com). If multiple episodes are found, it returns a playlist--otherwise, just the one result.

### CODE REVIEW

File: yt_dlp/extractor/podchaser.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# coding: utf-8
+from __future__ import unicode_literals
+import json
+import re
+from ..utils import float_or_none, try_get, str_to_int, unified_timestamp, merge_dicts
+from ..compat import compat_str
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:6]: use str

File: yt_dlp/extractor/podchaser.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
+# coding: utf-8
+from __future__ import unicode_literals
+import json
+import re
+from ..utils import float_or_none, try_get, str_to_int, unified_timestamp, merge_dicts
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py]: split lines and sort

File: yt_dlp/extractor/podchaser.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
+# coding: utf-8
+from __future__ import unicode_literals
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:2]: remove

File: yt_dlp/extractor/podchaser.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+import json
+import re
+from ..utils import float_or_none, try_get, str_to_int, unified_timestamp, merge_dicts
+from ..compat import compat_str
+from .common import InfoExtractor
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:7]: move above ..utils

File: yt_dlp/extractor/podchaser.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+
+class PodchaserIE(InfoExtractor):
+    _VALID_URL = r'''(?x)
+        https?://(?:www\.)?podchaser\.com/
+        (?:
+            (?:podcasts/[\w-]+-(?P<podcast_id>[\d]+)))
+        (?:/episodes/[\w\-]+-
+            (?P<id>[\d]+))?'''
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:16]: some of the brackets and indentation don't make sense

File: yt_dlp/extractor/podchaser.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    _TESTS = [{
+        'url': 'https://www.podchaser.com/podcasts/cum-town-36924/episodes/ep-285-freeze-me-off-104365585',
+        'info_dict': {
+            'id': '104365585',
+            'title': "Ep. 285 – freeze me off",
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:22]: use single quotes everywhere

File: yt_dlp/extractor/podchaser.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            'id': '699349',
+            'title': "Sean Carroll's Mindscape: Science, Society, Philosophy, Culture, Arts, and Ideas",
+            'description': 'md5:2cbd8f4749891a84dc8235342e0b5ff1'
+        },
+        'playlist_mincount': 225
+    }]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        audio_id, podcast_id = mobj.group('id'), mobj.group('podcast_id')
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:54]:
```suggestion
        audio_id, podcast_id = self._match_valid_url(url).group('id', 'podcast_id')
```

File: yt_dlp/extractor/podchaser.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+        # If one episode
+        if audio_id:
+            episodes = [self._download_json("https://api.podchaser.com/episodes/%s" % audio_id, audio_id)]
+
+        # Else get every episode available
+        else:
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:61]: This is better as 2 extractors imo

File: yt_dlp/extractor/podchaser.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            'description': episode.get('description'),
+            'url': episode.get('audio_url'),
+            'thumbnail': episode.get('image_url'),
+            'duration': str_to_int(episode.get('length')),
+            'timestamp': unified_timestamp(episode.get('air_date')),
+            'rating': float_or_none(episode.get('rating')),
+            'categories': [
+                x.get('text') for x in
+                podcast_data.get('categories')
+                or try_get(podcast_data, lambda x: x['summary']['categories'], list) or []],
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py]:
```suggestion
            'categories': traverse_obj(podcast_data, (('summary', None), 'categories', ..., 'text'))
```

File: yt_dlp/extractor/podchaser.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            'tags': [tag.get('text') for tag in podcast_data.get('tags') or []],
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:102]: similar to above

File: yt_dlp/extractor/podchaser.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+                        "start": i * 100,
+                        "count": 100,
+                        "sort_order": "SORT_ORDER_RECENT",
+                        "filters": {
+                            "podcast_id": podcast_id
+                        }, "options": {}
+                    }).encode())
+                curr_episodes = curr_episodes_data.get('entities') or []
+                if len(curr_episodes) + len(episodes) <= total_episode_count:
+                    episodes.extend(curr_episodes)
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:83]: should be a paged list

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/podchaser.py [hunk 12]
```diff
@@ -33,7 +31,34 @@ class PodchaserIE(InfoExtractor):
...
+    def _real_extract(self, url):
+        podcast_id, episode_id = self._match_valid_url(url).group('podcast_id', 'id')
+        podcast = self._download_json(self._API_BASE + 'podcasts/%s' % podcast_id, podcast_id)
+        episode = self._download_json(self._API_BASE + 'episodes/%s' % episode_id, episode_id)
+        return self._extract_episode(podcast, episode)
+
+
+class PodchaserFeedIE(PodchaserIE):
+    IE_NAME = 'Podchaser:feed'
+    _VALID_URL = PodchaserIE._VALID_URL_BASE + r'(?P<id>\d+)'
```
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/podchaser.py:60]: that doesn't quite work. Probably best to define a `suitable` instead:
```py
@classmethod
def suitable(cls, url):
    return super().suitable(url) and not PodchaserIE.suitable(url)
```

File: yt_dlp/extractor/podchaser.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            'start': page * self._PAGE_SIZE,
+            'count': self._PAGE_SIZE,
+            'sort_order': 'SORT_ORDER_RECENT',
+            'filters': {
+                'podcast_id': podcast_id
+            },
+            'options': {}
+        }
+        json_response = self._download_json(
+            self._API_BASE + 'list/episode', podcast_id,
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:91]: use fstrings instead of `+`. Same everywhere

File: yt_dlp/extractor/podchaser.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+            headers={'Content-Type': 'application/json;charset=utf-8'},
+            data=json.dumps(params).encode())
+        episodes = json_response.get('entities')
+        for episode in episodes:
+            yield self._extract_episode(podcast, episode)
+
+    def _real_extract(self, url):
+        podcast_id = self._match_valid_url(url).group('id')
```
Comment (pukkandan) [inline yt_dlp/extractor/podchaser.py:99]:
```suggestion
        podcast_id = self._match_id(url)
```

**Comment by connercsbn** [comment]
> I am starting to think that my suggestion to split the extractor may have been detrimental...
I'm happy to try whatever you think is best. Pretty new to OOP so this is a learning experience for me

**Pull request merged**